src/team_lib/mhba/xccl_mhba_team.c

```diff
@@ -400,30 +358,84 @@ xccl_status_t xccl_mhba_team_create_post(xccl_tl_context_t  *context,
         status = xccl_mhba_init_mkeys(mhba_team);
         if (status != XCCL_OK) {
             xccl_mhba_error("Failed to init mkeys");
-            goto remote_ctrl_fail;
+            goto malloc_fail;
+        }
+
+        // for each ASR - qp num, in addition to port lid, ctrl segment rkey and address, recieve mkey rkey
```

Comment:
as a TODO for probably another PR: this exchange of local_data now looks way to clumsy and complicated. I suggest we declare a local_data_t structure that would hold the fields: port, lid, rkey, addr, etc. Then pack and unpack after allgather will look nicer and cleaner.

Reply:
the problem is that that the rkeys size is dynamically allocated, based on the #nodes. i can't malloc and use ptr as it's need to be in same contiguous address for the allgather... how should i do it?

Reply:
local data has only 1 value for rkey - your local rkey. So, 
struct exchange_data {
port;
lid;
rkey;
addr;
}

struct exchange_data local_data;
local_data.port = port;
local_data.rkey = rkey;
...etc...

global_data = malloc(sizeof(struct exchange_data)*net_size);
allgather(&local_data, global_data, sizeof(exchange_data);

rkeys = malloc(sizeof(*rkeys)*net_size));
for (i =0; i<net_size; i++) {
rkeys[i] = global_data[i].rkey;
}

makes sense?

This is just to replace confusing pack/unpack into "local_data" array. Making it more clear. Can be part of other PR.

Reply:
sorry my bad - didnt mean rkey, but qp num. the amount of qp num is dynamic according to number of nodes.

Reply:
I would do this:
typedef struct exchange_data {
    int port;
    int lid;
    int rkey;
    void *addr;
   ...
   int qp_num[1];
} exchange_data_t;

size_t local_data_size = sizeof(exchange_data_t) + sizeof(int)*(net_size-1);
local_data = malloc(local_data_size);
global_data = malloc(local_data_size*local_size);

this way you can still access local_data->qp_num[i] but it resides contiguously in the memory as though it is part of the structure, so you can use allgather.